Updated iframe properties dialog

maufcost opened the pull request:
Fixes #1536

maufcost commented:
A sneak peak at our new iframe properties modal:

![iframe](https://user-images.githubusercontent.com/39862359/119672765-d9eaed80-be08-11eb-8d66-35bc19e492ee.gif)

jpeterson976 reviewed (changes requested):
This looks great! I especially love the sizing options, those SVGs look awesome. Great work!

One thing is that in the advanced options, the description text can be overlapped by the toggle button if the window isn't wide enough (shown below). 

<img width="562" alt="Screen Shot 2021-06-09 at 3 18 18 PM" src="https://user-images.githubusercontent.com/46502440/121419756-4e11af00-c93a-11eb-9a4c-f58ff99c6ca8.png">

Also, this may just be a me thing, but I think that the labels in the 'Options' section should be vertically centered with respect to their inputs as well. 

This is probably pretty nitpicky, but I think the 'Content type' label should be aligned with the options box - currently it's aligned with the container that contains the description text so it's sitting slightly below where it seems like it should. Similarly I think the 'Sizing' label should be aligned with the middle bullet point, though getting them to properly line up there might be annoying. (If anyone disagrees with this, I'm open to being convinced otherwise)

zachberry reviewed (changes requested):
Super excited about this, sorry for the delay in the review. I think you have the base of the form working pretty well, but I do have a bit of feedback on changes:

First screen:

* The preview input needs a debounce. If I try to type in a URL as soon as I input one character the iframe attempts to load right away (1)
* If you give the input an invalid URL, it will instead load the editor inside the iframe. Likewise, checking if the input is an iframe embed code could be simplified by trying to put the users' text as innerHTML inside a div and poking into the div to see if the browser created a valid iframe. You can see an example of both of these techniques inside the "parse-youtube-url.js" file, in the `parseYouTubeURL` function. It might be worthwhile to extract this code and generalize it to be able to be used in both places. (2)
* Add a 1px $color-shadow border to the iframe preview (3)
* The "Does the preview look good?" help-text state expands the dialog to suddenly be very wide. Want to make sure the width doesn't change between these states. (4)

Second screen:

* The Content type and Sizing boxes need `cursor:pointer` and a hover state. (5)
* The "px" in the Fixed size sizing option as well as the "!" in the warning icon under the Loading select box is using some default-level font. If these are in the SVG you'll want to use Libre Franklin (free to download) and then export the SVG with the fonts rendered as shapes. (6)
* Dialog needs a min-width so that the helper text below the Content type boxes doesn't wrap (7)
* Switching the different Content Type choices should set the options below automatically. By default 'Video or other media' would set Fit to Scale, Reload turned on and the other controls turned off. Embedded webpage should set Fit to Scroll and enable all controls. Additionally Sizing should default to Responsive - text width. (8)
* The text "Load when student sees IFrame" should be changed to "Load when student clicks IFrame" (9)

General:

* Buttons and inputs seem to be using Arial - want to use the standard Selects, Buttons and inputs for these. The Materia dialog uses SettingsDialogForm, that could be of use - It helps build out forms - Ian added it in the Materia branch. (10)
* When displayed in the viewer there seems to be a gray background to iframes - If a page doesn't define a background color it's more noticeable. (11)
* Editing an iframe should jump to the second screen if the iframe has a URL set (12)

maufcost commented:
I numbered the issues on Zac's last comment so they are easier to address and talk about.

**(1)** In order to add a debounce with our current library (`use-debounce`), we would need to transform our `EditIFrameModal` from a class-based component to a function component in order to work with the debounce hook. Is that ok if the `NewIFrameModal` is still a class-based component and the `EditIFrameModal` is a function one? Or do you prefer me to transform both of them to function components?
**(2)** I relied on the `parseYouTubeURL` to check for valid iframe embed codes and urls and everything works fine now, but the function turned out to be a lot simpler than the one for the youtube node. So, instead of joining them together, I was thinking about moving that file up some lex levels, and make both the iframe and youtube chunks import from the same file. Any thoughts?
**(3)** Done.
**(4)** Done.
**(5)** Done. I just need some feedback on the purple borders. Any thoughts about keeping the purple borders all the time or only on hover?
**(6)** Done.
**(7)** Done.
**(8)** Done.
**(9)** Done.
**(10)** Done. Applied the mixins and adapted inputs and selects to have the same border color.
**(11)** I haven't found any background colors set on iframes in the viewer. Could you send me any screenshots? :)
**(12)** Done.
**(13)** Proposing this one: I thought about disabling the button to go from the first modal to the second if the `src` typed is not valid. Any thoughts?

maufcost commented:
Finished applying the changes Zac requested. This PR is testable/reviewable again.

The tests that are not passing are some uncovered React functions I had to write in order to control more complicated hover/click actions. I'm just waiting to get your feedback on the current hover/click states I implemented before writing them.

jpeterson976 commented:
Looks good so far! Only thing I noticed is that setting it to "Embedded webpage" leaves out the Border setting when setting everything to true. Everything else seems to be working as expected with my testing. Good work!

For your proposal with (13), I think that it could be useful since it'd help prevent accidentally giving an invalid src and not realizing it. I don't think there are any cases where an invalid src would be intentional, but if they do exist then an extra warning message similar to the missing src message might be a better option.

Also, #1875 was recently merged in and removes the leading comma from the `controls` strings, so those should be safe to remove in the code now.

maufcost commented:
Thanks for all the suggestions so far everyone!

New changes include:
* Addressed the last changes that @jpeterson976  requested.
* I also found a handful of bugs with how `content types` and `controls` were handled. For that, I merged the `border` property into the `controls` property, allowing me to refactor a lot of code and better handle how the entire `controls` state is managed.
* I introduced a new property to the iframe context, called `controlsChanged` to better address Zac's **n#8** issue above. I was facing a bit of a boolean dilemma setting the controls to have specific properties depending on their iframe's content type _(because we would also need to know when users created iframes, turned off all controls, and closed them; created iframes, left the controls with their default options, and closed them; and all kinds of combinations of these user actions)_. This new prop solved this issue (apparently).

Let me know your thoughts :)

jpeterson976 reviewed (approved):
LGTM! Did some playing around with changing the settings, everything looks to be working as expected. Great work!

FrenjaminBanklin reviewed (dismissed):
After merging in changes from dev/31 and updating snapshots, this looks like it's running well.

FrenjaminBanklin reviewed (approved):
Merged in the current dev branch and checked functionality, it looks like all of this is working really well.

I did also contribute to this so it should probably have another person besides me look it over before I finalize it.

RachelDau commented:
These iframe options are really cool. I didn't find any issues with anything. I tested on different browsers and played around with the sizing and other settings.
